File: performers/java/sdk-performer-java/src/main/java/com/couchbase/sdk/utils/ClusterConnection.java
package com.couchbase.sdk.utils;

import com.couchbase.client.java.Cluster;
import com.couchbase.client.java.Collection;
import com.couchbase.grpc.sdk.protocol.ClusterConnectionCreateRequest;
import com.couchbase.grpc.sdk.protocol.DocLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;

/**
 * Creates a connection to the given cluster
 */
public class ClusterConnection {
    private final ClusterConnectionCreateRequest request;
    private static Cluster cluster;
    private static final Logger logger = LoggerFactory.getLogger(ClusterConnection.class);


    public ClusterConnection(ClusterConnectionCreateRequest reqData)  {
        this.request = reqData;
        var hostname = "couchbase://" + reqData.getClusterHostname();
        logger.info("Attempting connection to cluster " + hostname);
        cluster = Cluster.connect(hostname, reqData.getClusterUsername(), reqData.getClusterPassword());
        cluster.waitUntilReady(Duration.ofSeconds(30));
    }

    public Collection collection(DocLocation loc) {
        com.couchbase.grpc.sdk.protocol.Collection coll = null;

        if (loc.hasPool()) {
            coll = loc.getPool().getCollection();
        }
        else if (loc.hasSpecific()) {
            coll = loc.getSpecific().getCollection();
        }
        else if (loc.hasUuid()) {
            coll = loc.getUuid().getCollection();
        }
        else {
            throw new UnsupportedOperationException("Unknown DocLocation type");
        }

        var bucket = cluster.bucket(coll.getBucket());
        return bucket
                .scope(coll.getScope())
                .collection(coll.getCollection());
    }
}

File: performers/jvm/kotlin/src/main/kotlin/com/couchbase/client/performer/kotlin/util/ClusterConnection.kt
package com.couchbase.client.performer.kotlin.util

import com.couchbase.client.java.Bucket
import com.couchbase.client.kotlin.Cluster
import com.couchbase.client.performer.grpc.ClusterConnectionCreateRequest
import com.couchbase.client.performer.grpc.DocLocation
import java.util.concurrent.ConcurrentHashMap

class ClusterConnection(req: ClusterConnectionCreateRequest) {
    private val hostname = "couchbase://" + req.clusterHostname
    private val cluster = Cluster.connect(hostname, req.clusterUsername, req.clusterPassword)
    private val bucketCache: MutableMap<String, com.couchbase.client.kotlin.Bucket> = mutableMapOf()

    fun collection(loc: DocLocation): com.couchbase.client.kotlin.Collection {
        var coll: com.couchbase.client.performer.grpc.Collection? = null
        if (loc.hasPool()) coll = loc.pool.collection
        else if (loc.hasSpecific()) coll = loc.specific.collection
        else if (loc.hasUuid()) coll = loc.uuid.collection
        else throw UnsupportedOperationException("Unknown DocLocation type")

        // KCBC-98: no SDK bucket caching
        val bucket = bucketCache.getOrPut(coll.bucket) { cluster.bucket(coll.bucket) }
        return bucket.scope(coll.scope).collection(coll.collection)
    }
}